sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala

```diff
@@ -249,11 +261,28 @@ abstract class DataFrameActionImpl extends ActionSubFeedsImpl[DataFrameSubFeed]
     if (breakDataFrameLineage || preparedSubFeed.isStreaming.contains(true) || preparedSubFeed.filter.isDefined) preparedSubFeed = preparedSubFeed.breakLineage
     // enrich with fresh DataFrame if needed
     preparedSubFeed = enrichSubFeedDataFrame(input, preparedSubFeed, context.phase, isRecursive)
+    // add observations on input DataFrame
+    if (Environment.enableInputDataObjectCount) {
+      input match {
+        case evDataObject: ExpectationValidation => preparedSubFeed.dataFrame.foreach { df =>
+          // collect additional aggregate expressions for this DataObject from Action expectations with scope Job
+          val inputJobAggExpressionColumns = expectations.filter(_.scope == ExpectationScope.Job).flatMap(_.getInputAggExpressionColumns(id))
+          val additionalJobAggExpressionColumns = inputJobAggExpressionColumns.filter(c => c.getName.exists(n => n.endsWith(s"#${input.id.id}"))).map(c => c.as(c.getName.get.stripSuffix(s"#${input.id.id}"))) ++
+            inputJobAggExpressionColumns.filter(c => c.getName.exists(n => !n.contains("#") && input.id == prioritizedMainInputCandidates.head.id))
+          // for global input, validate constraints and expectations, otherwise just add default expectations, e.g. count
+          val isGlobalInputOnly = context.instanceRegistry.isInputOnlyDataObject(subFeed.dataObjectId)
```

Comment:
```suggestion
          val isInputOnlyDataObject = context.instanceRegistry.isInputOnlyDataObject(subFeed.dataObjectId)
```

what do you mean with globalInput? Do you mean that it's the input of the whole Action and not like the input of the second transformation in case there are mutlitple transformations of the action?

Reply:
You're right, naming improved.